### ISSUE TITLE
output of size [3000,3000]

### ISSUE THREAD

**Issue by TheNewSound** [gpujs/gpu.js#406]
Hi guys! Thanks for the awesome work on gpu.js

**_TLDR: WebGL loses context (crashes) when the output of the DICE / kernel function is gonna be more than approx. 512x512_**

I have a question, I've implemented the DICE similarity algorithm on gpu.js, it looks like the following:

```
const gpu = new GPU();
function DICE(strings, offset) {
    const y = this.thread.y+offset;
    const x = this.thread.x;
    if (x < y) {
        return 0;
    }
    const firstLength = strings[y][0];
    const secondLength = strings[x][0];
    const firstLengthA = strings[y][1];
    const secondLengthA = strings[x][1];
    if (firstLength === 0 && secondLength === 0)
        return 1;
    if (firstLength === 0 || secondLength === 0)
        return 0;
    /*if (strings[y] === strings[x])
     return 1;*/ // create dedicated function?
    if (firstLength === 1 && secondLength === 1)
        return 0;         // both are 1-letter strings
    if (firstLength < 2 || secondLength < 2)
        return 0;

    let intersectionSize = 0;
    for (let i = 1; i < secondLengthA; i++) {
        for (let j = 1; j < firstLengthA; j++) {
            if (strings[x][i * 2] === strings[y][j * 2]) { // bigram matches
                const val = Math.min(strings[x][(2 * i) + 1], strings[y][(2 * j) + 1]);
                intersectionSize = intersectionSize + val;
                break;
            }
        }
    }
    
    return (2.0 * intersectionSize) / (firstLength + secondLength - 2);
}
```
What it does, it calculates the similarity between every bigramArray-element, creating an upper-triangle matrix with similarity values.

Now, I call this function like this:

```
const myFunc = gpu.createKernel(DICE, {
      output: {x: bigramArray.length, y: bigramArray.length},
});

let result = myFunc(bigramArray, 0);
```

See that I call my DICE function with offset 0 (begin with the first row of bigramArray).
Where bigramArray is a 2d array, for example:

```
let bigramArray = [
[87, 68, 10343, 1, 10578, 1, 12141, 1, 25701, 3, 26469, 1, 26478, 4, 26723, 1, 26946, 1, 26988, 1, 26990, 2, 26994, 1, 27493, 2, …],

[87, 68, 10343, 1, 10578, 1, 12141, 1, ...],

...]
```
as you see, each element has the following format `[info, info, bigram, occurences, bigram, occurences, ...]`

The size of bigramArray can be around 3000 to 20000 records (with sub-arrays of approx 500 records).

Anyways, the problem: WebGL loses context (crashes) when the output of the DICE / kernel function is gonna be more than approx. 512x512
By the way, GPU.js runs in a web worker.

Should I batch-process smaller output pieces of the DICE / kernel function? **what is the best way to do that**?

**Comment by robertleeplummerjr** [comment]
Is it the same problem outlined here with a code same in js?: 
https://en.m.wikibooks.org/wiki/Algorithm_Implementation/Strings/Dice%27s_coefficient

The reason I ask is that generally I've seen you operations used to create a large output, which u see done in your algorithm, however in the one listed I see the output looks like a single float.  This sets off alarms in my mind because the GPU will be processing much of the same data for each fragment, which may actually result in something slower than the CPU.

**Comment by TheNewSound** [comment]
> Is it the same problem outlined here with a code same in js?:
> https://en.m.wikibooks.org/wiki/Algorithm_Implementation/Strings/Dice%27s_coefficient
> 
> The reason I ask is that generally I've seen you operations used to create a large output, which u see done in your algorithm, however in the one listed I see the output looks like a single float. This sets off alarms in my mind because the GPU will be processing much of the same data for each fragment, which may actually result in something slower than the CPU.

I implemented the algorithm in such a way that it won't process much of the same data for each fragment. I precalculated the bigrams (CPU does this really fast).
The only part of the DICE algorithm that is being done on the GPU is calculating the upper triangle similarity values (using precalculated bigram arrays).

I already have a performance increase of around 2 times vs an optimized-CPU only DICE implementation.

I was thinking about writing the results of a batch to a texture, and merging textures eventually. something like this (but I can't get it 100% working, still something not right, but I cant figure out what):

```
const texture = importAsTexture(bigramArray);
importAsTexture.destroy();
let result = gpuDICE(texture, 0);

for(let i = batchSize; i < bigramArray.length; i+=batchSize){
    const result2 = gpuDICE(texture,i);
    if(i+batchSize >= bigramArray.length-batchSize){//bigramArray.length){
        result = gpuConcat(result,result2,true);
    }else{
        result = gpuConcat(result,result2,false);
    }
}
```

where gpuConcat is defined as the following:

```
function gpuConcat(texture1, texture2, isLast){
    if(texture1.size[0]!==texture2.size[0]){
        return false;
    }
    const mergeTextures = gpu.createKernel(function(texture1,texture2,texture1size) {
        if(this.thread.y >= texture1size){
            return texture2[this.thread.y-texture1size][this.thread.x];
        }
        return texture1[this.thread.y][this.thread.x];
    })
     .setOutput({y: texture1.size[1]+texture2.size[1], x: texture1.size[0]})
     .setOutputToTexture(!isLast);
     let result = mergeTextures(texture1, texture2, texture1.size[1]);
     return result;
}

```

**Comment by robertleeplummerjr** [comment]
Are there any warnings?  It seems two things that catch my eye are the output size is generally defined as an array `[x,y,z]`, but I see an object used (at least in a brief skim over), and the output, when dynamic will likely need the `loopMaxIterations` to be set to the upper range or use a hardcoded value it constant.

**Comment by robertleeplummerjr** [comment]
Found thus article that isn't directly related, but a cool read about the dice algorithm nonetheless: https://medium.com/datadriveninvestor/deep-learning-in-medical-imaging-3c1008431aaf

**Comment by TheNewSound** [comment]
> Are there any warnings? It seems two things that catch my eye are the output size is generally defined as an array `[x,y,z]`, but I see an object used (at least in a brief skim over), and the output, when dynamic will likely need the `loopMaxIterations` to be set to the upper range or use a hardcoded value it constant.

I indeed got some warnings about "loopMaxIterations set to default 1000 ... should be defined as constant", but the code runs fine, when calculating an upper triangle matrix, the loop of elements to compare to get smaller as y gets bigger (down to the bottom there are a lot of zeros padded (0,0,0,0,0,0,0,0,0,0,0.9))

**Comment by robertleeplummerjr** [comment]
Can you increase it as a unit test?  After looking at your algorithm, I believe you may be able to write it to perform a few times faster.

**Comment by robertleeplummerjr** [comment]
>Increase it as a unit test? what do you mean? what should I increase?

I think I was typing on my phone, and meant to say can you  write it as a failing unit test.  Sorry for the typo.

`Uint16Array` Isn't officially supported yet.  Try and keep `Float32Array` used, that should help.

Closing as possibly self diagnosed.
